File: src/translation/tools/search_glossary.rs
use crate::llm::ToolDef;

#[derive(serde::Deserialize)]
pub struct Args {
    pub query: String,
}

pub fn def() -> ToolDef {
    ToolDef::new(
        "search_glossary",
        "Search the persistent glossary for canonical translations.\n\n\
            Behavior:\n\
            - Search BEFORE inventing a translation for names, titles, places, \
            organizations, abilities, items, or recurring terms.\n\
            - Batch multiple search_glossary() calls in one message.\n\
            - If no useful entry found, translate naturally and continue.\n\n\
            When to use: proper nouns or recurring terms likely to appear again.\n\
            When NOT to use: ordinary words or one-off everyday phrases.",
        serde_json::json!({
            "type": "object",
            "required": ["query"],
            "additionalProperties": false,
            "properties": {
                "query": {
                    "type": "string",
                    "description": "Source-language term to look up"
                }
            }
        }),
    )
    .strict()
}

File: typoon/translation/tools/search_glossary.py
from __future__ import annotations

from pydantic import BaseModel, Field

from typoon.llm.tool_dec import tool


class SearchGlossaryArgs(BaseModel):
    query: str = Field(description="Source-language term to look up")


@tool(strict=True)
async def search_glossary(args: SearchGlossaryArgs) -> str:
    """Search the persistent glossary for canonical translations.

    Only for terms NOT already listed in the user message glossary.
    Batch multiple calls in one message.
    If no match found, translate naturally and continue.

    When to use: proper nouns or recurring terms likely to appear again.
    When NOT to use: ordinary words, one-off phrases, or terms already in glossary.
    """
    raise NotImplementedError("dispatch handles this")
